hudi-common/src/main/java/org/apache/hudi/internal/schema/utils/AvroSchemaEvolutionUtils.java

```diff
@@ -111,17 +111,21 @@ public static InternalSchema reconcileSchema(Schema incomingSchema, InternalSche
     return SchemaChangeUtils.applyTableChanges2Schema(internalSchemaAfterAddColumns, typeChange);
   }
 
+  public static Schema reconcileSchema(Schema incomingSchema, Schema oldTableSchema) {
+    return convert(reconcileSchema(incomingSchema, convert(oldTableSchema)), oldTableSchema.getFullName());
+  }
+
   /**
-   * Reconciles nullability requirements b/w {@code source} and {@code target} schemas,
+   * Reconciles nullability and datatype requirements b/w {@code source} and {@code target} schemas,
    * by adjusting these of the {@code source} schema to be in-line with the ones of the
    * {@code target} one
    *
    * @param sourceSchema source schema that needs reconciliation
    * @param targetSchema target schema that source schema will be reconciled against
    * @param opts config options
-   * @return schema (based off {@code source} one) that has nullability constraints reconciled
+   * @return schema (based off {@code source} one) that has nullability constraints and datatypes reconciled
    */
-  public static Schema reconcileNullability(Schema sourceSchema, Schema targetSchema, Map<String, String> opts) {
+  public static Schema reconcileSchemaRequirements(Schema sourceSchema, Schema targetSchema, Map<String, String> opts) {
```

Comment:
Do we have unit testing on this?

Reply:
for now the end to end test cases cover all the scenarios

Reply:
both of  reconcileSchemaRequirements/reconcileSchema have the logical for type promotion
 Is it possible to merge these two methods

Reply:
@xiarixiaoyao - I agree there is scope to pull some common code out between the 2 apis, thinking it will be better that this can be handled in https://issues.apache.org/jira/browse/HUDI-6947 when we do the overall clean up. wdyt ?